src/Microsoft.Diagnostic.Repl/Command/CommandProcessor.cs

```diff
@@ -242,12 +290,67 @@ private void SetProperties(InvocationContext context, object instance)
                     property.Property.SetValue(instance, value);
                 }
 
-                if (_argument != null)
+                if (context != null && _argument != null)
                 {
-                    object value = context.ParseResult.CommandResult.GetValueOrDefault();
+                    object value = null;
+                    ArgumentResult result = context.ParseResult.CommandResult.ArgumentResult;
+                    switch (result)
+                    {
+                        case SuccessfulArgumentResult successful:
+                            value = successful.Value;
+                            break;
+                        case FailedArgumentResult failed:
+                            throw new InvalidOperationException(failed.ErrorMessage);
+                    }
                     _argument.SetValue(instance, value);
                 }
             }
+
+            private object[] BuildArguments(MethodBase methodBase, InvocationContext context)
+            {
+                ParameterInfo[] parameters = methodBase.GetParameters();
+                object[] arguments = new object[parameters.Length];
+                for (int i = 0; i < parameters.Length; i++) {
+                    Type parameterType = parameters[i].ParameterType;
+                    TryGetService(parameterType, context, out arguments[i]);
+                }
+                return arguments;
+            }
+
+            private bool TryGetService(Type type, InvocationContext context, out object service)
+            {
+                if (type == typeof(InvocationContext)) {
+                    service = context;
+                }
+                else if (!_commandProcessor._services.TryGetValue(type, out service)) {
+                    service = null;
+                    return false;
+                }
+                return true;
+            }
+        }
+
+        class LocalHelpBuilder : IHelpBuilder
+        {
+            private readonly CommandProcessor _commandProcessor;
+            private readonly HelpBuilder _helpBuilder;
+
+            public LocalHelpBuilder(CommandProcessor commandProcessor)
+            {
+                _commandProcessor = commandProcessor;
+                _helpBuilder = new HelpBuilder(commandProcessor.GetService<IConsole>(), maxWidth: Console.WindowWidth);
+            }
+
+            void IHelpBuilder.Write(ICommand command)
+            {
+                if (_commandProcessor._commandHandlers.TryGetValue(command.Name, out Handler handler))
+                {
+                    if (handler.InvokeHelp()) {
+                        return;
+                    }
+                }
```

Comment:
Is this so SOS help bubbles up? And does this preserve column width in both cases?

Reply:
This invokes the SOS dynamic help, but it doesn't honor the column width since it is the native SOS code printing this help out.